// ep_add_buttons/static/js/add_buttons.js
/*** 
*
* Most of the logic for task lists are done here
*
*/


if(typeof exports == 'undefined'){
  	var exports = this['mymodule'] = {};
}

/*** Add button to the editbar and bind a listener ***/
exports.add_buttons = {

	// Write the button to the dom
	init: function(){ 

		// make a sidebar with the button
		var buttonHTML = '<div id="sidebar">' + 
			'<div id="sidebarInner">' +
				'<button class="acl-write" id="analyseButton">' + 
					'<a class="grouped-right" data-l10n-id="pad.toolbar.add_buttons.title" title="Task list Checkbox">' + 
						'<span class="buttonicon buttonicon-add_buttons"></span> Analyseer Tekst' + 
					'</a>' + 
				'</button>' + 
			'</div>' + 
		'</div>';

		// append the sidebar and button to the pad content
		$('#editorcontainerbox').append(buttonHTML);

		// add an on click action to the button
		$('#analyseButton').click(function(){
			exports.add_buttons.onButton();
		});
		
		addDropdown();

	},	
	
	// execute this action on click	
	onButton: function(){ 
			
	analyzePadContent('click');
	},
}


exports.postAceInit = function(hook, context){
	startConnection();
	exports.add_buttons.init();
	
	var $head = $('[name="ace_outer"]').contents().find('[name="ace_inner"]').contents().find("head");                
	$head.append($("<link/>", 
    	{ rel: "stylesheet", href: "../static/plugins/ep_add_buttons/static/css/pad.css", type: "text/css" 			}));
}
